shards/sched.go

```diff
@@ -0,0 +1,542 @@
+func (s *multiScheduler) Acquire(ctx context.Context) (*process, error) {
+	if err := s.mu.RLock(ctx); err != nil {
+		return nil, err
+	}
+
+	// Start in interactive. yieldFunc will switch us to batch. sem can be nil
+	// if we fail while switching to batch. nil value prevents us releasing
+	// twice.
+	sem := s.semInteractive
+
+	if err := sem.Acquire(ctx); err != nil {
+		s.mu.RUnlock()
+		return nil, err
+	}
+
+	return &process{
+		releaseFunc: func() {
+			if sem != nil {
+				sem.Release()
+				sem = nil
+			}
+			s.mu.RUnlock()
+		},
+		yieldTimer: newDeadlineTimer(time.Now().Add(s.interactiveDuration)),
+		yieldFunc: func(ctx context.Context) error {
+			if sem != nil {
+				sem.Release()
+				sem = nil
+			}
+
+			// Try to acquire batch. Only set sem if we succeed so we know we can
+			// clean it up. If this fails we assume the process will stop running
+			// (ctx has expired).
+			semNext := s.semBatch
+			if err := semNext.Acquire(ctx); err != nil {
+				return err
+			}
+
+			sem = semNext
+			return nil
+		},
+	}, nil
+}
+
+// Exclusive implements scheduler.Exclusive.
+func (s *multiScheduler) Exclusive() *process {
+	// Exclusive process holds a write lock on mu, which ensures we have no
+	// processes running (search semaphores are empty).
+	//
+	// exclusive processes will never yield, so we leave yieldTimer and
+	// yieldFunc nil.
+	s.mu.Lock()
+	return &process{
+		releaseFunc: func() {
+			s.mu.Unlock()
+		},
+	}
+}
+
+// semaphoreScheduler shares a single semaphore for all searches. An exclusive
+// process acquires the full semaphore. This is equivalent to how concurrency
+// is managed in upstream. It exists as a fallback while we test
+// multiScheduler.
+type semaphoreScheduler struct {
+	throttle *semaphore.Weighted
+	capacity int64
+}
+
+// Acquire implements scheduler.Acquire.
+func (s *semaphoreScheduler) Acquire(ctx context.Context) (*process, error) {
+	return s.acquire(ctx, 1)
+}
+
+// Exclusive implements scheduler.Exclusive.
+func (s *semaphoreScheduler) Exclusive() *process {
+	// Won't error since context.Background won't expire.
+	proc, _ := s.acquire(context.Background(), s.capacity)
+	return proc
+}
+
+func (s *semaphoreScheduler) acquire(ctx context.Context, weight int64) (*process, error) {
+	if err := s.throttle.Acquire(ctx, weight); err != nil {
+		return nil, err
+	}
+	return &process{
+		releaseFunc: func() {
+			s.throttle.Release(weight)
+		},
+	}, nil
+}
+
+// process represents a running search query or an exclusive process. When the
+// process is done a call to Release is required.
+type process struct {
+	// yieldTimer ensures we only call yieldFunc once after a deadline.
+	yieldTimer *deadlineTimer
+	// yieldFunc is called once by Yield.
+	yieldFunc func(context.Context) error
+
+	// releaseFunc is called once by Release
+	releaseFunc func()
+}
+
+// Release the resources/locks/semaphores associated with this process. Can
+// only be called once.
+func (p *process) Release() {
```

Comment:
Should we define an interface for `process`? I think it might be nice to use an interface to clarify how we define a process. Just like for the scheduler. Nothing we have to do now, of course.

Reply:
there is only one implementation of process though? What benefit does an interface provide?

Reply:
Generally speaking, with one implementation, I agree that the added value is none. But, in this context, process is an important conceptual entity (for the lack of a better word) just like `scheduler` and it could be nicely defined by its behavior instead of a concrete struct. This is totally subjective of course.